### ISSUE TITLE
i18n Guidelines

### ISSUE THREAD

**Issue by MikaelSmith** [puppetlabs/cppbestpractices#10]
Use Leatherman logging, locale, i18n support.

Use http://www.boost.org/doc/libs/1_56_0/libs/locale/doc/html/collation.html for sorting.

**Comment by MikaelSmith** [comment]
Document uses of Boost.Locale: sorting, multiple pluralization, masculine/feminine, casing, splitting strings (by words, separators, white-space).

Translate individual strings or smaller.

What do we use utfcpp for?

**Comment by MikaelSmith** [comment]
TODO: update Facter fact output sorting to use locale. Custom facts can introduce any character and need to be sorted correctly.

**Comment by MikaelSmith** [comment]
Blocked on https://github.com/puppetlabs/leatherman/pull/99.

**Comment by kylog** [comment]
@MikaelSmith or @ahenroid: today's lunch-and-learn reminded me. Are our i18n best practices for c++ ready to be captured here?

**Comment by ahenroid** [comment]
@kylog, yes, as far as I'm aware, C++ i18n best practices have been demonstrated by the cpp-project-template, but really not documented here or anywhere else internally. I'm happy to take a first pass at this, unless @MikaelSmith, you have this written down elsewhere and/or have lots of detail in your head that you want to get down first.

**Comment by ahenroid** [comment]
And I'm thinking specifically of string externalization. There is already some info around UTF-8.

**Comment by MikaelSmith** [comment]
I think I've pointed you at everything I previously wrote.
